src/TestRunner.lua

```diff
@@ -55,30 +48,27 @@ function TestRunner.runPlanNode(session, planNode, lifecycleHooks)
 
 		messagePrefix = messagePrefix or ""
 
-		local testEnvironment = getfenv(callback)
-
-		for key, value in pairs(TestRunner.environment) do
+		local originalEnvironment = getfenv(callback)
+		local testEnvironment = setmetatable({}, { __index = originalEnvironment })
+		for key, value in pairs(planNode.environment) do
 			testEnvironment[key] = value
 		end
+		setfenv(callback, testEnvironment)
 
-		testEnvironment.fail = function(message)
-			if message == nil then
-				message = "fail() was called."
-			end
+		local nodeSuccess, nodeResult = xpcall(callback, function(message)
+			return debug.traceback(message, 2)
```

Comment:
This is interesting. Why do we want to start our traceback capture at the second level of the stack?

Reply:
Unless I messed up my testing, 1 is the innermost error handler and 2 is the user's code. (Should probably add an automated test case for that.)

Reply:
Oh, but just passing in debug.traceback with no wrapper also works. Will fix.